sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -0,0 +1,1107 @@
+    public enum ContentStructure {
+      /** If the content structure is not specified, the default value BUNDLE will be used. */
+      CONTENT_STRUCTURE_UNSPECIFIED,
+      /**
+       * The source file contains one or more lines of newline-delimited JSON (ndjson). Each line is
+       * a bundle, which contains one or more resources. Set the bundle type to history to import
+       * resource versions.
+       */
+      BUNDLE,
+      /**
+       * The source file contains one or more lines of newline-delimited JSON (ndjson). Each line is
+       * a single resource.
+       */
+      RESOURCE,
+      /** The entire file is one JSON bundle. The JSON can span multiple lines. */
+      BUNDLE_PRETTY,
+      /** The entire file is one JSON resource. The JSON can span multiple lines. */
+      RESOURCE_PRETTY
+    }
+  }
+
+  /** The type Execute bundles. */
+  public static class ExecuteBundles extends PTransform<PCollection<String>, Write.Result> {
+    private final String fhirStore;
+
+    /**
+     * Instantiates a new Execute bundles.
+     *
+     * @param fhirStore the fhir store
+     */
+    ExecuteBundles(ValueProvider<String> fhirStore) {
+      this.fhirStore = fhirStore.get();
+    }
+
+    /**
+     * Instantiates a new Execute bundles.
+     *
+     * @param fhirStore the fhir store
+     */
+    ExecuteBundles(String fhirStore) {
+      this.fhirStore = fhirStore;
+    }
+
+    @Override
+    public FhirIO.Write.Result expand(PCollection<String> input) {
+      return Write.Result.in(
+          input.getPipeline(),
+          input
+              .apply(ParDo.of(new ExecuteBundlesFn(fhirStore)))
+              .setCoder(HealthcareIOErrorCoder.of(StringUtf8Coder.of())));
+    }
+
+    /** The type Write Fhir fn. */
+    static class ExecuteBundlesFn extends DoFn<String, HealthcareIOError<String>> {
+
+      private Counter failedBundles = Metrics.counter(ExecuteBundlesFn.class, "failed-bundles");
+      private transient HealthcareApiClient client;
+      private final ObjectMapper mapper = new ObjectMapper();
+      /** The Fhir store. */
+      private final String fhirStore;
+
+      /**
+       * Instantiates a new Write Fhir fn.
+       *
+       * @param fhirStore the Fhir store
+       */
+      ExecuteBundlesFn(String fhirStore) {
+        this.fhirStore = fhirStore;
+      }
+
+      /**
+       * Initialize healthcare client.
+       *
+       * @throws IOException the io exception
+       */
+      @Setup
+      public void initClient() throws IOException {
+        this.client = new HttpHealthcareApiClient();
+      }
+
+      /**
+       * Execute Bundles.
+       *
+       * @param context the context
+       */
+      @ProcessElement
+      public void executeBundles(ProcessContext context) {
+        String body = context.element();
+        try {
+          // Validate that data was set to valid JSON.
+          mapper.readTree(body);
+          client.executeFhirBundle(fhirStore, body);
```

Comment:
This is executing a single RPC per element, right? This will likely be slow, and lead to underutilization of worker resources. As a follow up, messages could be buffered, and RPCs issued via a thread pool

Reply:
IMO Batching should be handled upstream from the connector based on what type of bundles the user is trying to commit to fire.

executeBundle is a single call per (beam) element (but each (beam) element in this case is already a [FHIR Bundle](https://www.hl7.org/fhir/bundle.html)) but in a transaction type bundle it's very crucial what elements are batched together (as they are committed as a single transaction on the back-end). The user must control this to get the desired behavior.

with batch type bundle the operations are executed individually on the back end (but provided by the client as a single call). 
We could provide a convenience PTransform for buffering batches into a bundle. 

context of what we support in Google HCAPI:
https://cloud.google.com/healthcare/docs/reference/rest/v1beta1/projects.locations.datasets.fhirStores.fhir/executeBundle

broader context of what the FHIR spec has for bundle type:
https://www.hl7.org/fhir/valueset-bundle-type.html

Reply:
To be a bit clearer:

- FHIR bundle is 1:1 with beam element here (but may contain many operations). 
- The FHIR bundle a field that dictates if it is transaction or batch type but they use the same endpoint / method.

Reply:
Makes sense. Sounds good to me!